lib/spack/spack/spec.py

```diff
@@ -2781,8 +2855,13 @@ def ensure_no_deprecated(root):
     def _new_concretize(self, tests=False):
         import spack.solver.asp
 
-        if not self.name:
-            raise spack.error.SpecError("Spec has no name; cannot concretize an anonymous spec")
+        self.replace_hash()
+
+        for node in self.traverse():
+            if not node.name:
+                raise spack.error.SpecError(
+                    f"Spec {node} has no name; cannot concretize an anonymous spec"
+                )
```

Comment:
I guess we don't currently model this, but I don't think it's unreasonable to concretize, say, `hdf5` with *some* dependency built with gcc, e.g.: `hdf5 ^%gcc`.  `develop` fails much less gracefully on this, though:

```console
> spack spec zlib ^%gcc
==> Error: Invalid module name:
```

vs. this branch:

```console
> spack spec zlib ^%gcc
==> Error: Spec %gcc has no name; cannot concretize an anonymous spec
```

So I guess this is an improvement and we can "fix" it later if needed.